analytics_data_api/v0/models.py

```diff
@@ -497,6 +497,44 @@ def get_simple_data_for_all_students(self, course_id):
         return ModuleEngagement.objects.all().filter(course_id=course_id) \
             .order_by('username')
 
+    def get_aggregate_engagement_data(self, course_id):
+        seven_days_ago = datetime.datetime.today() - datetime.timedelta(days=7)
+        query = ModuleEngagement.objects.filter(course_id=course_id)\
+            .values('username')\
+            .annotate(
+                videos_overall=Sum(Case(When(entity_type='video', then='count'), output_field=IntegerField()))) \
+            .annotate(
+                videos_last_week=Sum(Case(When(entity_type='video', created__gt=seven_days_ago, then=1),
+                                          output_field=IntegerField()))) \
+            .annotate(
+                problems_overall=Sum(Case(When(entity_type='problem', then='count'), output_field=IntegerField()))) \
+            .annotate(
+                problems_last_week=Sum(Case(When(entity_type='problem', created__gt=seven_days_ago, then='count'),
+                                            output_field=IntegerField()))) \
+            .annotate(
+                correct_problems_overall=Sum(Case(When(entity_type='problem', event='completed', then='count'),
+                                                  output_field=IntegerField()))) \
+            .annotate(
+                correct_problems_last_week=Sum(Case(When(entity_type='problem', event='completed',
+                                                         created__gt=seven_days_ago, then='count'),
+                                                    output_field=IntegerField()))) \
+            .annotate(
+                problems_attempts_overall=Sum(Case(When(entity_type='problem', event='attempted', then='count'),
+                                                   output_field=IntegerField()))) \
+            .annotate(
+                problems_attempts_last_week=Sum(Case(When(entity_type='problem', event='attempted',
+                                                          created__gt=seven_days_ago, then='count'),
+                                                     output_field=IntegerField()))) \
+            .annotate(
+                forum_posts_overall=Sum(Case(When(entity_type='discussion', then='count'),
+                                             output_field=IntegerField()))) \
+            .annotate(
+                forum_posts_last_week=Sum(Case(When(entity_type='discussion', created__gt=seven_days_ago, then='count'),
+                                               output_field=IntegerField()))) \
+            .annotate(
```

Comment:
I think annotate can take a single dict, which would let us reduce this down to one call & reduce some grossness.

Also I feel like we probably don't have linting running in CI if this is building green. Could you check this with `make quality` after `make requirements` (you may want to do this in a virtualenv)

Reply:
linting is running. I ran make quality and everything was green.

Reply:
Here is the SQL query generated - it appears to be a standard aggregate. Not sure what else can be done to optimize it:
SELECT module_engagement.username, SUM(CASE WHEN module_engagement.entity_type = 'video' THEN count ELSE NULL END) AS videos_overall, 
SUM(CASE WHEN (module_engagement.created > STR_TO_DATE('2019-08-07', '%Y-%m-%d') AND module_engagement.entity_type = 'video') THEN count ELSE NULL END) 
AS videos_last_week, 
SUM(CASE WHEN module_engagement.entity_type = 'problem' THEN count ELSE NULL END) AS problems_overall, 
SUM(CASE WHEN (module_engagement.created > STR_TO_DATE('2019-08-07', '%Y-%m-%d') AND module_engagement.entity_type = 'problem') THEN count ELSE NULL END) 
AS problems_last_week, 
SUM(CASE WHEN (module_engagement.event = 'completed' AND module_engagement.entity_type = 'problem') THEN count ELSE NULL END) 
AS correct_problems_overall, 
SUM(CASE WHEN (module_engagement.created > STR_TO_DATE('2019-08-07', '%Y-%m-%d') AND module_engagement.event = 'completed' 
AND module_engagement.entity_type = 'problem') THEN module_engagement.count ELSE NULL END) 
AS correct_problems_last_week, 
SUM(CASE WHEN (module_engagement.event = 'attempted' AND module_engagement.entity_type = 'problem') THEN count ELSE NULL END) AS problems_attempts_overall, 
SUM(CASE WHEN (module_engagement.created > STR_TO_DATE('2019-08-07', '%Y-%m-%d') AND module_engagement.event = 'attempted'
AND module_engagement.entity_type = 'problem') THEN count ELSE NULL END) AS problems_attempts_last_week, 
SUM(CASE WHEN module_engagement.entity_type = 'discussion' THEN count ELSE NULL END) AS forum_posts_overall, 
SUM(CASE WHEN (module_engagement.created > STR_TO_DATE('2019-08-07', '%Y-%m-%d') AND module_engagement.entity_type = 'discussion') THEN count ELSE NULL END) 
AS forum_posts_last_week, MAX(module_engagement.created) AS date_last_active 
FROM module_engagement
WHERE module_engagement.course_id = 'course-v1:seakademieX+1+2T2018'
GROUP BY module_engagement.username;

Reply:
The EXPLAIN info we got from this (out of band) was actually surprisingly uninstructive. Marking as resolved.

Reply:
Or, wait, the core of this comment was that we should switch to object-arg'd `annotate`. Let's still do that

Reply:
I am having difficulty tracking down object-arg'd annotate. Can you link to a doc or existing code base.

Reply:
Based purely on the fact that there are stars in front of the kwargs in https://docs.djangoproject.com/en/2.2/ref/models/querysets/#django.db.models.query.QuerySet.annotate, I suspect you can provide any number of kwargs to a single call to annotate, which is really what I'm after here.

Reply:
I concur: a single `annotate()` call should be nicer.